lib/coins/btc.js

```diff
@@ -34,23 +41,30 @@ function formatAddress (address) {
   return address
 }
 
-function validate (network, address) {
+function bech32Validator (network, address) {
+  let decoded
   try {
-    if (!network) throw new Error('No network supplied.')
-    if (!address) throw new Error('No address supplied.')
-
-    const buf = bs58check.decode(address)
-    const addressType = buf[0]
-
-    if (buf.length !== 21) throw new Error(`Invalid length: ${buf.length}`)
-
-    if (network === 'main' && _.includes(addressType, [0x00, 0x05])) return true
-    if (network === 'test' && _.includes(addressType, [0x6f, 0xc4])) return true
+    decoded = bech32.decode(address)
+  } catch (error) {
+    console.log('Failed to decode bech32 address')
+    return false
+  }
 
-    throw new Error('General error')
-  } catch (err) {
-    console.log(err)
-    console.log('Invalid bitcoin address: [%s] %s', network, address)
+  const witnessVersion = decoded.words[0]
+  if (witnessVersion !== 0) {
+    console.log('Unsupported witness version for bech32')
     return false
   }
+
```

Comment:
@kokipsiho I think you misunderstood my comment. I just wanted the length checking code to be outside the if/else block, I didn't want to get rid of it completely. You probably wondered why I would want to get rid of the check. If so, please ask about things like this, or we risk losing important checks. I'm sure there will be a lot of things I don't catch in the future, so we both have to be looking at this carefully.
 
```javascript
  const data = bech32.fromWords(decoded.words.slice(1))	
  if (data.length !== 20 && data.length !== 32) {	
    console.log(`Invalid bech32 address length: ${data.length}`)	
    return false	
  }
```

Reply:
I think what confused me is that length is determined by witness version. In future I will ask when in doubt.

Reply:
👍